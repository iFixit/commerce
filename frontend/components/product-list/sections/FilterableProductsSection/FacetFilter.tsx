import { getRefinementDisplayType } from '@helpers/product-list-helpers';
import { ProductList } from '@models/product-list';
import * as React from 'react';
import { RefinementMultiSelect } from './RefinementMultiSelect';
import { RefinementSingleSelect } from './RefinementSingleSelect';
import { RefinementDisplayType } from '@models/product-list/types';

type FacetFilterProps = {
   attribute: string;
   productList: ProductList;
   onClose?: () => void;
};

export function FacetFilter({
   attribute,
   productList,
   onClose,
}: FacetFilterProps) {
   switch (getRefinementDisplayType(attribute)) {
      case RefinementDisplayType.SingleSelect:
         return (
            <RefinementSingleSelect
               attribute={attribute}
               showMore
               showMoreLimit={200}
               productList={productList}
               onClose={onClose}
            />
         );
      case RefinementDisplayType.MultiSelect:
         return (
            <RefinementMultiSelect
               attribute={attribute}
               showMore
               showMoreLimit={200}
            />
         );
      default:
         throw new Error('Unknown refinement display type');
   }
}
