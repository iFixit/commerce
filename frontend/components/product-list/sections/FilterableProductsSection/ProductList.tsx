import placeholderImageUrl from '@assets/images/no-image-fixie.jpeg';
import {
   Badge,
   Box,
   Button,
   Divider,
   Flex,
   Heading,
   HStack,
   LinkBox,
   LinkOverlay,
   Stack,
   Text,
   VStack,
} from '@chakra-ui/react';
import { ThemeTypings } from '@chakra-ui/styled-system';
import { Rating } from '@components/ui';
import { useAppContext } from '@ifixit/app';
import { ProductSearchHit } from '@models/product-list';
import { IfixitImage } from '@components/ifixit-image';
import * as React from 'react';
import { useProductSearchHitPricing } from './useProductSearchHitPricing';

export type ProductListProps = React.PropsWithChildren<unknown>;

export function ProductList({ children }: ProductListProps) {
   return (
      <VStack
         data-testid="list-view-products"
         spacing={2}
         align="stretch"
         width="full"
         boxSizing="border-box"
         divider={<Divider mt="0" style={{ marginTop: 0, marginBottom: 0 }} />}
      >
         {children}
      </VStack>
   );
}

export interface ProductListItemProps {
   product: ProductSearchHit;
}

export function ProductListItem({ product }: ProductListItemProps) {
   const appContext = useAppContext();

   const { price, compareAtPrice, isDiscounted, percentage } =
      useProductSearchHitPricing(product);

   const productHeadingId = `product-heading-${product.handle}`;

   const quantityAvailable = Math.max(0, product.quantity_available);

   const showProBadge = product.is_pro > 0;
   const showDiscountBadge = quantityAvailable > 0 && isDiscounted;
   const showLifetimeWarrantyBadge =
      quantityAvailable > 0 && product.lifetime_warranty;
   const showOemPartnershipBadge =
      quantityAvailable > 0 && product.oem_partnership;
   const showBadges =
      showProBadge ||
      showDiscountBadge ||
      showLifetimeWarrantyBadge ||
      showOemPartnershipBadge;

   return (
      <LinkBox as="article" aria-labelledby={productHeadingId} role="group">
         <Flex
            align="flex-start"
            py={{
               base: 4,
               xl: 7,
            }}
            px="4"
         >
            <Box
               flexGrow={0}
               flexShrink={0}
               w={{
                  base: '100px',
                  sm: '160px',
                  md: '140px',
                  lg: '160px',
               }}
            >
               {product.image_url ? (
                  <IfixitImage
                     src={product.image_url}
                     alt={product.title}
                     objectFit="contain"
                     width="180px"
                     height="180px"
                  />
               ) : (
                  <IfixitImage
                     src={placeholderImageUrl}
                     alt={product.title}
                     sizes="180px"
                  />
               )}
            </Box>
            <Stack
               direction={{
                  base: 'column',
                  xl: 'row',
               }}
               spacing={{
                  base: 1,
                  xl: 2,
               }}
               flexGrow={1}
               ml="4"
            >
               <Box flexShrink={1} w="full">
                  <Heading
                     id={productHeadingId}
                     as="h3"
                     fontSize={{
                        base: 'md',
                        lg: 'lg',
                     }}
                     mb={{
                        base: 2,
                        xl: 4,
                     }}
                     _groupHover={{ color: 'brand.500' }}
                  >
                     {product.title}
                  </Heading>
                  <Text
                     noOfLines={3}
                     fontSize={{
                        base: 'sm',
                        sm: 'sm',
                        lg: 'md',
                     }}
                     mb={{
                        base: 2,
                        xl: 4,
                     }}
                  >
                     {product.short_description}
                  </Text>
                  {(product.rating >= 4 || product.rating_count > 10) && (
                     <HStack align="center" data-testid="reviewStars">
                        <Rating value={product.rating} />
                        <Text
                           fontSize={{
                              base: 'sm',
                              lg: 'md',
                           }}
                        >
                           {product.rating_count}
                        </Text>
                     </HStack>
                  )}
                  {showBadges && (
                     <Flex
                        wrap="wrap"
                        mt="2"
                        sx={{
                           '& > *': {
                              mr: 1,
                              mt: 1,
                           },
                        }}
                     >
                        {showProBadge &&
                           ProductListItemBadge({
                              content: 'iFixit Pro',
                              colorScheme: 'orange',
                           })}
                        {showDiscountBadge &&
                           ProductListItemBadge({
                              content: `${percentage}% Off`,
                              colorScheme: 'red',
                           })}
                        {showOemPartnershipBadge &&
                           ProductListItemBadge({
                              content: product.oem_partnership!,
                              colorScheme: 'green',
                           })}
                        {showLifetimeWarrantyBadge &&
                           ProductListItemBadge({
                              content: 'Lifetime Warranty',
                              colorScheme: 'blue',
                           })}
                     </Flex>
                  )}
               </Box>
               <VStack
                  flexShrink={0}
                  align="flex-end"
                  alignSelf={{
                     base: 'flex-end',
                     xl: 'flex-start',
                  }}
               >
                  {price != null && (
                     <VStack
                        align="flex-end"
                        spacing="0"
                        mt={{
                           base: 2,
                           sm: 0,
                           md: 2,
                           lg: 0,
                        }}
                     >
                        <Text
                           color={isDiscounted ? 'red.700' : 'inherit'}
                           fontWeight="bold"
                           fontSize="xl"
                           lineHeight="1em"
                           data-testid="product-price"
                        >
                           ${price}
                        </Text>
                        {isDiscounted && (
                           <Text
                              lineHeight="1em"
                              textDecoration="line-through"
                              color="gray.400"
                              data-testid="product-compared-at-price"
                           >
                              ${compareAtPrice}
                           </Text>
                        )}
                     </VStack>
                  )}
                  <Stack
                     direction={{
                        base: 'row-reverse',
                        sm: 'column',
                        md: 'row-reverse',
                        lg: 'column',
                     }}
                     align={{
                        base: 'center',
                        sm: 'flex-end',
                        md: 'center',
                        lg: 'flex-end',
                     }}
                     spacing={{
                        base: 3,
                        sm: 2,
                     }}
                  >
                     <LinkOverlay
                        href={`${appContext.ifixitOrigin}${product.url}`}
                     >
                        <Button
                           as="div"
                           minW={{
                              base: '100px',
                              sm: '60px',
                              md: '100px',
                              lg: '60px',
                           }}
                           colorScheme="brand"
                        >
                           View
                        </Button>
                     </LinkOverlay>
                     {quantityAvailable < 10 && quantityAvailable > 0 && (
                        <Text color="gray.500" fontSize="14px">
                           Only {quantityAvailable} left in stock
                        </Text>
                     )}
                  </Stack>
               </VStack>
            </Stack>
         </Flex>
      </LinkBox>
   );
}

interface ProductListItemBadgeProps {
   content: string;
   colorScheme: ThemeTypings['colorSchemes'];
}

function ProductListItemBadge({
   content,
   colorScheme,
}: ProductListItemBadgeProps) {
   return (
      <Badge
         colorScheme={colorScheme}
         textTransform="none"
         borderRadius="lg"
         px="2.5"
         py="1"
      >
         {content}
      </Badge>
   );
}
