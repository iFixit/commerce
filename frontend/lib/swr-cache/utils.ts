import isPlainObject from 'lodash/isPlainObject';
import { z } from 'zod';

export interface CacheAdapter {
   name?: string;
   set(key: string, value: CacheEntry): Promise<void>;
   get(key: string): Promise<CacheEntry | null>;
   delete(key: string): Promise<void>;
}

const CacheEntrySchema = z.object({
   metadata: z.object({
      ttl: z.number().positive().optional(),
      staleWhileRevalidate: z.number().optional(),
      createdAt: z.number(),
   }),
   value: z.unknown(),
});

type CacheEntry = z.infer<typeof CacheEntrySchema>;

interface CreateCacheEntryOptions {
   ttl?: number;
   staleWhileRevalidate?: number;
}

export const createCacheEntry = <T = unknown>(
   value: T,
   options: CreateCacheEntryOptions
): CacheEntry => {
   const { ttl, staleWhileRevalidate } = options;
   return {
      metadata: {
         ttl,
         staleWhileRevalidate: ensurePositiveNumber(staleWhileRevalidate),
         createdAt: Date.now(),
      },
      value,
   };
};

export const isValidCacheEntry = (entry: unknown): entry is CacheEntry => {
   const validation = CacheEntrySchema.safeParse(entry);
   if (!validation.success) {
      return false;
   }
   return isFresh(validation.data) || isStale(validation.data);
};

export const isStale = (entry: CacheEntry) => {
   const now = Date.now();
   const {
      ttl = Infinity,
      staleWhileRevalidate = 0,
      createdAt,
   } = entry.metadata;

   return !isFresh(entry) && createdAt + ttl + staleWhileRevalidate > now;
};

export const createCacheKey = (keyPrefix: string, variables: unknown) => {
   return `${keyPrefix}:${stableHash(variables)}`;
};

export const sleep = (ms: number) =>
   new Promise((resolve) => setTimeout(resolve, ms));

export const ensurePositiveNumber = (value: number | null | undefined) => {
   return value == null ? 0 : Math.max(0, value);
};

const isFresh = (entry: CacheEntry) => {
   const now = Date.now();
   const { ttl = Infinity, createdAt } = entry.metadata;
   return createdAt + ttl > now;
};

const stableHash = (variables: unknown): string => {
   return JSON.stringify(variables, (_, val) =>
      isPlainObject(val)
         ? Object.keys(val)
              .sort()
              .reduce((result, key) => {
                 result[key] = val[key];
                 return result;
              }, {} as any)
         : val
   );
};
