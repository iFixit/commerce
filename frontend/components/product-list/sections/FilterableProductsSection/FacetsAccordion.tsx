import {
   Accordion,
   AccordionButton,
   AccordionIcon,
   AccordionItem,
   AccordionItemProps,
   AccordionPanel,
   Box,
   forwardRef,
   HStack,
   Text,
   VStack,
} from '@chakra-ui/react';
import { formatFacetName } from '@helpers/algolia-helpers';
import { ProductList } from '@models/product-list';
import * as React from 'react';
import { useHits } from 'react-instantsearch-hooks-web';
import { FacetFilter } from './FacetFilter';
import { useCountRefinements } from './useCountRefinements';
import { MAX_VALUES_PER_FACET, useFilteredFacets } from './useFacets';
import { useFilteredRefinementList } from './useFilteredRefinementList';

type FacetsAccordianProps = {
   productList: ProductList;
};

const initialExpandedFacets = ['facet_tags.Item Type'];

export function FacetsAccordion({ productList }: FacetsAccordianProps) {
   const facets = useFilteredFacets(productList);
   const countRefinements = useCountRefinements();
   const [indexes, setIndexes] = React.useState<number[]>(() => {
      return initialExpandedFacets
         .map((expandedFacet) =>
            facets.findIndex((facet) => facet === expandedFacet)
         )
         .filter((index) => index >= 0);
   });

   const handleChangeIndexes = React.useCallback((indexes: number[]) => {
      setIndexes(indexes);
   }, []);

   return (
      <Accordion
         allowMultiple
         index={indexes}
         onChange={handleChangeIndexes}
         data-testid="facets-accordion"
         sx={{
            '> .visible': {
               borderTopWidth: '0px',
            },
            '> .visible ~ .visible': {
               borderTopWidth: '1px',
            },
            '>:last-child': {
               borderBottom: 'none',
            },
         }}
      >
         {facets.map((facet, facetIndex) => {
            const facetAttributes = [facet];
            if (facet === 'price_range') {
               facetAttributes.push('facet_tags.Price');
            }
            const refinedCount = countRefinements(facetAttributes);
            return (
               <FacetAccordionItem
                  key={facet}
                  attribute={facet}
                  refinedCount={refinedCount}
                  productList={productList}
                  isExpanded={indexes.includes(facetIndex)}
               />
            );
         })}
      </Accordion>
   );
}

type FacetAccordionItemProps = AccordionItemProps & {
   attribute: string;
   refinedCount: number;
   productList: ProductList;
   isExpanded: boolean;
};

export const FacetAccordionItem = forwardRef<FacetAccordionItemProps, 'div'>(
   ({ attribute, refinedCount, productList, isExpanded, ...props }, ref) => {
      const { items } = useFilteredRefinementList({
         attribute,
         limit: MAX_VALUES_PER_FACET,
      });
      const { hits } = useHits();
      const isProductListEmpty = hits.length === 0;
      const hasApplicableRefinements = items.length > 0;
      const isDisabled = isProductListEmpty || !hasApplicableRefinements;

      const formattedFacetName = formatFacetName(attribute);
      const isHidden = !hasApplicableRefinements && !isProductListEmpty;

      return (
         <AccordionItem
            ref={ref}
            {...props}
            hidden={isHidden}
            isDisabled={isDisabled}
            data-testid={`${
               isExpanded ? 'expanded' : 'collapsed'
            }-facet-accordion-item-${attribute}`}
            data-facet-name={attribute}
            className={isHidden ? 'hidden' : 'visible'}
         >
            <AccordionButton
               aria-label={
                  isExpanded
                     ? `Collapse ${formattedFacetName}`
                     : `Expand ${formattedFacetName}`
               }
            >
               <Box flex="1" textAlign="left" fontWeight="bold">
                  {formattedFacetName}
               </Box>
               <HStack>
                  {refinedCount > 0 && (
                     <Text
                        rounded="full"
                        bg="gray.600"
                        color="white"
                        px="1.5"
                        fontSize="xs"
                     >
                        {refinedCount}
                     </Text>
                  )}
                  <AccordionIcon />
               </HStack>
            </AccordionButton>
            <AccordionPanel pb={4} display={isDisabled ? 'none' : 'block'}>
               <VStack align="stretch" spacing="3">
                  <FacetFilter
                     attribute={attribute}
                     productList={productList}
                  />
               </VStack>
            </AccordionPanel>
         </AccordionItem>
      );
   }
);
