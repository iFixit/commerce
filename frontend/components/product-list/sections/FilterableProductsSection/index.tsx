import {
   ProductListEmptyStateIllustration,
   SearchEmptyStateIllustration,
} from '@assets/svg';
import {
   Box,
   BoxProps,
   Button,
   Flex,
   forwardRef,
   Heading,
   HStack,
   Icon,
   Link,
   Text,
   VStack,
} from '@chakra-ui/react';
import { Card } from '@components/ui';
import { IFIXIT_ORIGIN } from '@config/env';
import { getProductListTitle } from '@helpers/product-list-helpers';
import { cypressWindowLog } from '@helpers/test-helpers';
import { useLocalPreference } from '@ifixit/ui';
import {
   ProductList as TProductList,
   ProductSearchHit,
} from '@models/product-list';
import * as React from 'react';
import {
   useClearRefinements,
   useCurrentRefinements,
   useHits,
   useSearchBox,
} from 'react-instantsearch-hooks-web';
import { CurrentRefinements } from './CurrentRefinements';
import { FacetsAccordion } from './FacetsAccordion';
import { Pagination } from './Pagination';
import { ProductGrid, ProductGridItem } from './ProductGrid';
import { ProductList, ProductListItem } from './ProductList';
import { ProductViewType, Toolbar } from './Toolbar';
import { useDevicePartsItemType } from './useDevicePartsItemType';

const PRODUCT_VIEW_TYPE_STORAGE_KEY = 'productViewType';

type SectionProps = {
   productList: TProductList;
};

export function FilterableProductsSection({ productList }: SectionProps) {
   const { hits } = useHits<ProductSearchHit>();
   const [viewType, setViewType] = useLocalPreference(
      PRODUCT_VIEW_TYPE_STORAGE_KEY,
      ProductViewType.List
   );

   cypressWindowLog({ filteredProducts: hits });

   const productsContainerScrollRef = useScrollIntoViewEffect([hits]);

   const isEmpty = hits.length === 0;

   return (
      <Flex
         ref={productsContainerScrollRef}
         as="section"
         direction="column"
         align="stretch"
         mb="4"
         data-testid="filterable-products-section"
         aria-labelledby="filterable-products-section-heading"
      >
         <Heading as="h2" id="filterable-products-section-heading" srOnly>
            Products
         </Heading>
         <Toolbar
            viewType={viewType}
            productList={productList}
            onViewTypeChange={setViewType}
         />
         <CurrentRefinements />
         <HStack mt="4" align="flex-start" spacing={{ base: 0, md: 4 }}>
            <FacetCard>
               <FacetsAccordion productList={productList} />
            </FacetCard>
            <Card flex={1}>
               <ProductListEmptyState
                  productList={productList}
                  hidden={!isEmpty}
               />
               {!isEmpty && viewType === ProductViewType.Grid && (
                  <ProductGrid>
                     {hits.map((hit) => {
                        return (
                           <ProductGridItem key={hit.handle} product={hit} />
                        );
                     })}
                  </ProductGrid>
               )}
               {!isEmpty && viewType === ProductViewType.List && (
                  <ProductList>
                     {hits.map((hit) => {
                        return (
                           <ProductListItem key={hit.objectID} product={hit} />
                        );
                     })}
                  </ProductList>
               )}
               <Pagination />
            </Card>
         </HStack>
      </Flex>
   );
}

type FacetCardProps = React.PropsWithChildren<{}>;

function FacetCard({ children }: FacetCardProps) {
   return (
      <Card
         py="1.5"
         px="3"
         w="250px"
         boxShadow="md"
         maxH="calc(100vh - var(--chakra-space-4) * 2)"
         overflow="auto"
         display={{
            base: 'none',
            md: 'block',
         }}
         position="sticky"
         top="4"
      >
         {children}
      </Card>
   );
}

function useScrollIntoViewEffect(deps: React.DependencyList) {
   const ref = React.useRef<HTMLDivElement>(null);
   React.useEffect(() => {
      if (ref.current && ref.current.getBoundingClientRect().top < 0) {
         ref.current.scrollIntoView({
            behavior: 'smooth',
         });
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
   }, deps);
   return ref;
}

type EmptyStateProps = BoxProps & {
   productList: TProductList;
};

const ProductListEmptyState = forwardRef<EmptyStateProps, 'div'>(
   ({ productList, ...otherProps }, ref) => {
      const clearRefinements = useClearRefinements();

      const currentRefinements = useCurrentRefinements();
      const hasRefinements = currentRefinements.items.length > 0;

      const searchBox = useSearchBox();
      const hasSearchQuery = searchBox.query.length > 0;

      const isFiltered = hasRefinements || hasSearchQuery;

      const itemType = useDevicePartsItemType(productList);
      const title = getProductListTitle(productList, itemType);
      const encodedQuery = encodeURIComponent(searchBox.query);

      if (isFiltered) {
         return (
            <VStack
               ref={ref}
               pt="16"
               pb="20"
               px="2"
               textAlign="center"
               {...otherProps}
            >
               <Icon
                  as={SearchEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text fontSize="lg" fontWeight="bold" w="full">
                  No matching products found in {title}
               </Text>
               <Text maxW="500px" color="gray.500">
                  Try adjusting your search or filter to find what you&apos;re
                  looking for.
               </Text>
               <Link
                  maxW="500px"
                  color="brand.500"
                  href={`${IFIXIT_ORIGIN}/Search?query=${encodedQuery}`}
               >
                  Search all of iFixit for&nbsp;
                  <Text as="span" fontWeight="bold">
                     {searchBox.query}
                  </Text>
               </Link>
               <Box pt="8">
                  <Button
                     colorScheme="brand"
                     onClick={() => {
                        searchBox.clear();
                        clearRefinements.refine();
                     }}
                  >
                     Reset filters
                  </Button>
               </Box>
            </VStack>
         );
      }
      return (
         <Card
            ref={ref}
            pt="16"
            pb="20"
            borderRadius={{ base: 'none', sm: 'lg' }}
            {...otherProps}
         >
            <VStack>
               <Icon
                  as={ProductListEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text
                  fontSize="lg"
                  fontWeight="bold"
                  w="full"
                  textAlign="center"
               >
                  Empty collection
               </Text>
               <Text maxW="500px" color="gray.500" textAlign="center" px="2">
                  This collection does not have products. Try to navigate to
                  subcategories to find what you are looking for.
               </Text>
            </VStack>
         </Card>
      );
   }
);
