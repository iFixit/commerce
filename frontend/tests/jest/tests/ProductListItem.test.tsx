import { renderWithAppContext } from '../utils';
import { screen } from '@testing-library/react';
import { ProductListItem } from '@templates/product-list/sections/FilterableProductsSection/ProductList';
import { mockedProductSearchHit } from '../__mocks__/products';

jest.mock(
   '@templates/product-list/sections/FilterableProductsSection/useProductSearchHitPricing',
   () => ({
      ...jest.requireActual(
         '@templates/product-list/sections/FilterableProductsSection/useProductSearchHitPricing'
      ),
      useProductSearchHitPricing: jest.fn(() => {
         return {
            price: { amount: 5, currencyCode: 'usd' },
            compareAtPrice: { amount: 5, currencyCode: 'usd' },
            isDiscounted: 5,
            percentage: 5,
         };
      }),
   })
);

jest.mock('@ifixit/ui/commerce/hooks/useUserPrice', () => ({
   ...jest.requireActual('@ifixit/ui/commerce/hooks/useUserPrice'),
   useUserPrice: jest.fn(() => {
      return {
         price: { amount: 5, currencyCode: 'usd' },
         compareAtPrice: { amount: 5, currencyCode: 'usd' },
      };
   }),
}));

describe('ProductListItem', () => {
   it('renders and matches the snapshot', () => {
      // @ts-ignore
      const { asFragment } = renderWithAppContext(
         <ProductListItem product={mockedProductSearchHit} />
      );
      const shortDescription = screen.getByText(
         'Replace a dead or malfunctioning model EB-BG96aasd5ABE battery in a Samsung Galaxy S9 Plus smartphone.'
      );

      (expect(shortDescription) as any).toBeInTheDocument();
      (expect(asFragment()) as any).toMatchSnapshot();
   });

   it('renders without the review stars', () => {
      // We don't render the stars if the rating <= 4 AND rating_count < 10
      const mockedProductSearchHitWithRating = {
         ...mockedProductSearchHit,
         rating: 3.5,
         rating_count: 9,
      };

      // @ts-ignore
      renderWithAppContext(
         <ProductListItem product={mockedProductSearchHitWithRating} />
      );

      const reviewStars = screen.queryByTestId('reviewStars');
      (expect(reviewStars) as any).not.toBeInTheDocument();
   });

   it('renders with the review stars', () => {
      // We render the stars if the rating >= 4 OR rating_count > 10

      // If rating < 4; count > 10
      const mockedProductSearchHitWithRating = {
         ...mockedProductSearchHit,
         rating: 1,
         rating_count: 15,
      };

      // @ts-ignore
      const { rerender } = renderWithAppContext(
         <ProductListItem product={mockedProductSearchHitWithRating} />
      );
      const reviewStars = screen.queryByTestId('reviewStars');
      (expect(reviewStars) as any).toBeInTheDocument();

      // If rating > 4; count < 10
      mockedProductSearchHitWithRating.rating = 5;
      mockedProductSearchHitWithRating.rating_count = 5;
      // @ts-ignore
      rerender(<ProductListItem product={mockedProductSearchHitWithRating} />);
      (expect(reviewStars) as any).toBeInTheDocument();

      // If rating > 4; count > 10
      mockedProductSearchHitWithRating.rating = 5;
      mockedProductSearchHitWithRating.rating_count = 15;
      // @ts-ignore
      rerender(<ProductListItem product={mockedProductSearchHitWithRating} />);
      (expect(reviewStars) as any).toBeInTheDocument();
   });
});
