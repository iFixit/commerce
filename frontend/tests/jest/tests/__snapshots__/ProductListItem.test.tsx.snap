// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListItem renders and matches the snapshot 1`] = `
<DocumentFragment>
  <article
    aria-labelledby="product-heading-galaxy-s9-plus-replacement-battery"
    class="chakra-linkbox css-10n1vic"
    role="group"
  >
    <div
      class="css-mnl840"
    >
      <div
        class="css-1vy26t"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27320%27%20height=%27320%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Galaxy S9+ Battery"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="chakra-stack css-1ryy1c1"
      >
        <div
          class="css-1c6uy1d"
        >
          <div
            class="chakra-stack css-16syxvt"
          >
            <div
              class="css-j7qwjs"
            >
              <h3
                class="chakra-heading css-10ht6wa"
                id="product-heading-galaxy-s9-plus-replacement-battery"
              >
                Galaxy S9+ Battery
              </h3>
              <p
                class="chakra-text css-a5k9pc"
              >
                Replace a dead or malfunctioning model EB-BG96aasd5ABE battery in a Samsung Galaxy S9 Plus smartphone.
              </p>
            </div>
            <div
              class="chakra-stack css-84zodg"
              data-testid="reviewStars"
            >
              <div
                class="chakra-stack css-134p2gk"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-star  css-86s11o"
                  data-icon="star"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-star  css-86s11o"
                  data-icon="star"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-star  css-86s11o"
                  data-icon="star"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-star  css-86s11o"
                  data-icon="star"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-star-half  css-1p6kd4g"
                  data-icon="star-half"
                  data-prefix="fad"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    class="fa-duotone-group"
                  >
                    <path
                      class="fa-secondary"
                      d="M146.3 512C145.3 512.1 144.2 512.1 143.1 512H146.3zM288 439.8V-.0387L288.1-.0391C300.4-.0391 311.6 6.954 316.9 17.97L381.2 150.3L524.9 171.5C536.8 173.2 546.8 181.6 550.6 193.1C554.4 204.7 551.3 217.3 542.7 225.9L438.5 328.1L463.1 474.7C465.1 486.7 460.1 498.9 450.2 506C440.3 513.1 427.2 514 416.5 508.3L288.1 439.8L288 439.8z"
                      fill="currentColor"
                    />
                    <path
                      class="fa-primary"
                      d="M288 439.8L159.8 508.3C149 514 135.9 513.1 126 506C116.1 498.9 111.1 486.7 113.2 474.7L137.8 328.1L33.58 225.9C24.97 217.3 21.91 204.7 25.69 193.1C29.46 181.6 39.43 173.2 51.42 171.5L195 150.3L259.4 17.97C264.7 6.995 275.8 .0131 287.1-.0391L288 439.8zM433.2 512C432.1 512.1 431 512.1 429.9 512H433.2z"
                      fill="currentColor"
                    />
                  </g>
                </svg>
              </div>
              <p
                class="chakra-text css-itvw0n"
              >
                18
              </p>
            </div>
            <div
              class="css-1qc3xp7"
            >
              <div
                class="css-rzte29"
              >
                <div
                  class="css-zvlevn"
                >
                  5% Off
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-stack css-f4fjr9"
        >
          <div
            class="css-oazic6"
          >
            <p
              class="chakra-text css-wxbpi4"
              data-testid="price"
            >
              $5.00
            </p>
          </div>
          <div
            class="chakra-stack css-15cy0v5"
          >
            <a
              class="chakra-linkbox__overlay css-1hnz6hu"
              href="/products/galaxy-s9-plus-replacement-battery"
            >
              <div
                class="chakra-button css-1sntzjs"
              >
                View
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
